orkaudio/filters/LiveStream/LiveStreamFilter.cpp

```diff
@@ -0,0 +1,265 @@
+	// to 16 bits internally.
+	// 0 = 8-bit samples
+	// 1 = 16-bit samples
+	char sound_size = 1;
+
+	// @param sound_type Mono or stereo sound
+	// 0 = Mono sound
+	// 1 = Stereo sound
+	//char sound_type = outputDetails.m_channel == 0 ? 0 : 1;
+	char sound_type = 1;
+
+	timestamp += 160; //Timestamp increment = clock frequency/frame rate
+					  //160 byte payload of G.711 has a packetization interval of 20 ms
+					  //For 1 second, there will be 1000ms / 20ms = 50 frames
+					  //Audio RTP packet timestamp incremental value = 8kHz / 50 = 8000Hz / 50 = 160
+
+	if (isFirstPacket)
+	{
+		headChannel = outputDetails.m_channel;
+		isFirstPacket = false;
+	}
+
+	if (outputDetails.m_channel == headChannel && status)
+	{
+		bufferQueue.push(inputBuffer);
+	}
+
+	if (rtmp != NULL && status)
+	{
+		if (outputDetails.m_channel != headChannel && bufferQueue.size() > 0)
+		{
+			char *outputBuffer = (char *)malloc(size);
+			char *tempBuffer = bufferQueue.front();
+			bufferQueue.pop();
+
+			for (int i = 0; i < 160; ++i)
+			{
+				outputBuffer[i * 2] = tempBuffer[i];
+				outputBuffer[i * 2 + 1] = inputBuffer[i];
+			}
+
+			if (srs_audio_write_raw_frame(rtmp, sound_format, sound_rate, sound_size, sound_type, outputBuffer, size, timestamp) != 0)
+			{
+				srs_human_trace("send audio raw data failed.");
+				return;
+			}
+
+			srs_human_trace("Accepted Packet: sample_rate=%d, sound_format=%s, m_timestamp=%d, m_arrivalTimestamp=%d,m_sequenceNumber=%d",
+							outputDetails.m_sampleRate, outputDetails.m_rtpPayloadType, outputDetails.m_timestamp, outputDetails.m_arrivalTimestamp, outputDetails.m_sequenceNumber);
+			srs_human_trace("sent packet: type=%s, time=%d, size=%d, codec=%d, rate=%d, sample=%d, channel=%d",
+							srs_human_flv_tag_type2string(SRS_RTMP_TYPE_AUDIO), timestamp, size, sound_format, sound_rate, sound_size, sound_type);
+		}
+	}
+}
+
+void LiveStreamFilter::AudioChunkOut(AudioChunkRef &chunk)
+{
+	chunk = m_outputAudioChunk;
+}
+
+AudioEncodingEnum LiveStreamFilter::GetInputAudioEncoding()
+{
+	return UnknownAudio;
+}
+
+AudioEncodingEnum LiveStreamFilter::GetOutputAudioEncoding()
+{
+	return UnknownAudio;
+}
+
+CStdString LiveStreamFilter::GetName()
+{
+	return "LiveStreamFilter";
+}
+
+bool LiveStreamFilter::SupportsInputRtpPayloadType(int rtpPayloadType)
+{
+	//so that BatchProcessing doesn't pick this filter.
+	return rtpPayloadType == pt_Unknown;
+}
+
+void LiveStreamFilter::CaptureEventIn(CaptureEventRef &event)
+{
+	//Start RTP Stream Open
+	auto key = event->EventTypeToString(event->m_type);
+	LOG4CXX_INFO(s_log, "LiveStream CaptureEventIn " + key + " : " + event->m_value);
+	if (event->m_type == CaptureEvent::EventTypeEnum::EtCallId)
+	{
+		m_callId = event->m_value;
+	}
+
+	if (event->m_type == CaptureEvent::EventTypeEnum::EtKeyValue && event->m_key == "LiveStream" && event->m_value == "start")
+	{
+		std::string url = "rtmp://" + CONFIG.m_rtmpServerEndpoint + ":" + CONFIG.m_rtmpServerPort + "/live/" + m_callId;
```

Comment:
what if `m_callId` is empty?

Reply:
@kingster there is a check before calling this function